util/src/main/java/io/zeebe/util/health/CriticalComponentsHealthMonitor.java

```diff
@@ -0,0 +1,120 @@
+  private final ActorControl actor;
+  private final Logger log;
+  private FailureListener failureListener;
+
+  public CriticalComponentsHealthMonitor(final ActorControl actor, final Logger log) {
+    this.actor = actor;
+    this.log = log;
+    this.failureListener = null;
+  }
+
+  public void startMonitoring() {
+    actor.runAtFixedRate(HEALTH_MONITORING_PERIOD, this::updateHealth);
+  }
+
+  @Override
+  public void monitorComponent(final String componentName) {
+    actor.run(() -> componentHealth.put(componentName, HealthStatus.UNHEALTHY));
+  }
+
+  @Override
+  public void removeComponent(final String componentName) {
+    actor.run(
+        () -> {
+          monitoredComponents.remove(componentName);
+          componentHealth.remove(componentName);
+        });
+  }
+
+  @Override
+  public void registerComponent(final String componentName, final HealthMonitorable component) {
+    actor.run(
+        () -> {
+          monitoredComponents.put(componentName, component);
+          component.addFailureListener(
+              FailureListener.withListeners(
+                  () -> onComponentFailure(componentName),
+                  () -> onComponentRecovered(componentName)));
+          componentHealth.put(componentName, component.getHealthStatus());
+        });
+  }
+
+  @Override
+  public HealthStatus getHealthStatus() {
+    return healthStatus.get();
+  }
+
+  @Override
+  public void addFailureListener(final FailureListener failureListener) {
+    actor.run(() -> this.failureListener = failureListener);
+  }
+
+  private void onComponentFailure(final String componentName) {
+    actor.call(
+        () -> {
+          log.error("{} failed, marking it as unhealthy", componentName);
+          componentHealth.put(componentName, HealthStatus.UNHEALTHY);
+          reCalculateHealth();
+        });
+  }
+
+  private void onComponentRecovered(final String componentName) {
+    actor.call(
+        () -> {
+          log.error("{} failed, marking it as unhealthy", componentName);
+          componentHealth.put(componentName, HealthStatus.UNHEALTHY);
+        });
+  }
+
+  private void updateHealth() {
+    componentHealth.keySet().stream()
+        .forEach(component -> componentHealth.put(component, getHealth(component)));
+    reCalculateHealth();
+  }
+
+  private void reCalculateHealth() {
+    final boolean healthy =
+        componentHealth.values().stream()
+            .allMatch(healthStatus -> healthStatus == HealthStatus.HEALTHY);
+    healthStatus.set(healthy ? HealthStatus.HEALTHY : HealthStatus.UNHEALTHY);
+    if (!healthy) {
+      notifyFailureListener();
+    }
+  }
+
+  private void notifyFailureListener() {
+    if (this.failureListener != null) {
+      failureListener.onFailure();
+    }
+  }
+
+  private HealthStatus getHealth(final String componentName) {
+    final HealthMonitorable component = monitoredComponents.get(componentName);
+    if (component != null) {
```

Comment:
Is this possible? It would be a bug if it happened, no? Then we probably want to log when it is actually null.

Reply:
You can add the name of the component to monitor before the component is available. For eg:- BrokerHealthCheckService knows about all partitions before the partitions are registered and ready to monitor.